### PR TITLE
React 16 Upgrade

### LINKED ISSUE #614: React 16
This issue will track React 16 compatibility. It seems to work fine but there are several peer dependency warnings.

Dependencies compatibility:

- [x] [React 16 final release](https://code.facebook.com/posts/1716776591680069/react-16-a-look-inside-an-api-compatible-rewrite-of-our-frontend-ui-library/)
- [x] [Enzyme 3.0](https://github.com/airbnb/enzyme/blob/master/CHANGELOG.md#300)
- [x] [markdown-to-jsx](https://github.com/probablyup/markdown-to-jsx) — N/A (seems to work)
- [x] [react-codemirror2](https://github.com/scniro/react-codemirror2) — ~~no support~~
- [x] [react-group](https://github.com/sapegin/react-group) — ~~N/A~~
- [x] [react-icons](https://github.com/gorangajic/react-icons) — ~~[no support]~~(https://github.com/gorangajic/react-icons/issues/103)
- [x] ~~[react-modal](https://github.com/reactjs/react-modal) — [some issues](https://github.com/reactjs/react-modal/issues/498) (we use it only in examples)~~

* **N/A** means no reported issues.
* **No support** means `peerDependencies` don’t allow React 16, may work otherwise.

Styleguidist upgrade status:

- [ ] Upgrade dev dependencies to React 16 — #631
- [x] Migrate to Enzyme 3.0 — #618
- [x] Remove react-addons-test-utils (not used anymore)
- [x] Test or update markdown-to-jsx
- [x] Update react-codemirror2
- [x] Test or update react-group — tested, works fine
- [x] Update react-icons
- [x] Update or remove react-modal
- [x] Add global [componentDidCatch handler](https://blog.bugsnag.com/react-16-error-handling/) — #627
- [x] Test Styleguidist with React 16: there should be no warning in browser console or npm warning on `npm install`

### PR DESCRIPTION
**Pull request by tizmagik**
Everything looks like it's working with no warnings or errors that I could see, except:

 - CRA example not working (also an issue on `master` I think?), filed a separate issue here: #630 
 - Test failures with Preview error handling test -- I would like to discuss this further...

For the Preview component, I think we might need to change up how we do error handling in React 16? I'm wondering if instead of a try/catch if we need to add a `componentDidCatch()` to the preview renderer and catch/report errors there. Since React is a peerDep, this would have the nice side affect of not triggering when users are running with React 15 and the old try/catch behavior would still work, I think.

Please have a look and let me know what you think we should do...

Closes #614

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,213 +1,215 @@
 {
```
Comment (sapegin) [inline package.json]: Indentation was changed for some reason. There should be two spaces.

**Comment by sapegin** [comment]
Yeah, sounds good, if it will work in React 15 and 16.

File: test/raf-polyfill.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// requestAnimationFrame “polyfill”
+window.requestAnimationFrame = a => a();
+global.requestAnimationFrame = window.requestAnimationFrame;
```
Comment (tizmagik) [inline test/raf-polyfill.js]: Note that this is technically an incorrect shim, it should be a `setTimeout()` or something to properly simulate async nature of rAF.

Comment (sapegin) [reply test/raf-polyfill.js]: I don’t care about async nature ;-)

Comment (tizmagik) [reply test/raf-polyfill.js]: Hm yea, that might be a slippery slope though, especially when async rendering is enabled by default in the future.

**Comment by tizmagik** [comment]
Does VSCode not respect editorconfig? Also not sure why it only does it for that package.json

**Comment by tizmagik** [comment]
Editorconfig was on. I don't know, I give up, I ran prettier manually via CLI and won't touch that file anymore 🤞

**Comment by sapegin** [comment]
Please don’t run Prettier on package.json — it’s all messed up now ;-|

**Comment by sapegin** [comment]
> Does VSCode not respect editorconfig? 

It has a plugin for that. But I have a feeling that it doesn’t always work.

**Comment by tizmagik** [comment]
Yea I had the plugin and it was enabled, not sure why it messed up the spacing... ah well that's not the important diff for right now, need to get the tests working

**Comment by tizmagik** [comment]
This upgrade might not be as straight forward actually, need to give how we're error handling some more thought...

image

Full message

```
Error: A cross-origin error was thrown. React doesn't have access to the actual error object in development. See https://fb.me/react-crossorigin-error for more information.
```
More info, https://fb.me/react-crossorigin-error

---

If anyone has any ideas, feel free to commit to this branch or branch off of it and run with it 👍

**Comment by sapegin** [comment]
As I understand we’re trying to print an error from another React root:

https://github.com/styleguidist/react-styleguidist/blob/master/src/rsg-components/Preview/Preview.js#L99-L103

Probably we should use componentDidCatch and rethrow it or something like that.

**Comment by sapegin** [comment]
@tizmagik Did you have time to investigate the issue? Do you need any help with this?

**Comment by tizmagik** [comment]
Haven’t had time to look at this since my last commit but hope to pick this up again this weekend. But, as always, feel free to jump in if you have any ideas!

**Comment by tizmagik** [comment]
Cool I think we're good now @sapegin , all synced up with master too

File: examples/basic/package.json [hunk 3]
```diff
@@ -25,9 +25,10 @@
     "css-loader": "^0.28.4",
     "dog-names": "^1.0.2",
     "lodash": "^4.17.4",
-    "prop-types": "^15.5.10",
-    "react": "^15.4.2",
-    "react-dom": "^15.4.2",
+    "prop-types": "^15.6.0",
+    "react": "^16.0.0",
+    "react-dom": "^16.0.0",
+    "react-modal": "^3.0.0",
```
Comment (sapegin) [inline examples/basic/package.json]: We’re not using react-modal anymore.

File: src/rsg-components/Preview/Preview.js [hunk 4]
```diff
@@ -90,7 +90,7 @@ export default class Preview extends Component {
 
 		const exampleComponent = this.evalInContext(compiledCode);
 		const wrappedComponent = (
-			<Wrapper>
+			<Wrapper onError={err => this.handleError(err)}>
```
Comment (sapegin) [inline src/rsg-components/Preview/Preview.js]: We’re using `bind` in constructor in other components.

Comment (tizmagik) [reply src/rsg-components/Preview/Preview.js]: Hm, I dunno why but the tests failed with the bind. Thoughts?

Comment (sapegin) [reply src/rsg-components/Preview/Preview.js]: You were adding it to another component — I’ve fixed it and it seems to work fine now ;-)

**Comment by sapegin** [comment]
Thanks, merged! 🍕

File: src/rsg-components/Preview/Preview.js (hunk 4 above)
Comment (tizmagik) [reply src/rsg-components/Preview/Preview.js]: Oh haha thanks, was context switching too much I guess 😁

**Pull request merged**